### ISSUE TITLE
JavaScript

### ISSUE THREAD

**Issue by mattblackdev** [tagtime/TagTime#74]
Has anyone worked on a JavaScript port for this? I could see it as a neat microservice on node and as a PWA or React Native app. Might be a fun side project to work on a port.

**Comment by dreeves** [comment]
I don't think so! I agree that would be wonderful. And reminder that the domain tagtime.com is still up for grabs for the first person to match all the features of the Perl+xterm+vim implementation (which @bsoule and I still use all day every day).

**Comment by mattblackdev** [comment]
@dreeves that's really cool to hear y'all still use it daily. I'd be interested to use it myself but I'd like to get pinged on my phone when I'm not on the computer. So, I might actually work on a ReactNative version.

**Comment by shepheb** [comment]
I'm actually working on an Android version using Firebase. (I work on Firebase at Google, and I'm building it as a Hackathon project this week.) That could easily be expanded to a web or iOS version.

That said, I don't have ownership of the concept, so go for it if you want to build a client. Doubly so if you wanted to build an iOS one!

**Comment by mattblackdev** [comment]
@shepheb very cool! Keep us updated!

**Comment by waynemcdougall** [comment]
I've set up a version that uses Twitter direct messaging. So I get notifications on my phone. Follow @cathleenbot for a taste. 

I'm not proud of the code, and it doesn't have all the features I want. But happy to share if anyone thinks it may be useful.

**Comment by mattblackdev** [comment]
That’s genius!

**Comment by waynemcdougall** [comment]
Well you can always rely on me to find the free option. I just needed something that gave notifications on my phone without having to write my own app. 

Of course the delivery mecanism need not be Twitter. Even email would work. But the notifications work nicely.

**Comment by Insti** [comment]
1) Is there much/any delay with twitter pings?
2) How do you respond? Also tweet replies?

**Comment by waynemcdougall** [comment]
1. I have not observed any delay except after power cuts. The outstanding pings are then sent, and that can trigger twitter rate limiting so if the power is out for 12 hours it may be 3 hours before all the pings come through. I time stamp them in the ping itself although not with seconds. So I can get some in the same minute. The backend database stores the full time of the ping. 

2. Yes also tweet replies but to be clear, these are direct messages so they are private. 

So I have one account, my usual Twitter account, which follows @cathleenbot. Cathy sends my account the pings as a direct message, which notifies me on my phone and PC. I reply in the direct message conversation. I used to use tags, but autocorrect keeps arguing so I'm sending to use whole words now. 

I did try one account, but you don't get notifications if you send a direct message to yourself. 

I didn't bother implementing the code that makes all pings at the same time for everyone, although easy to add. But, interestingly, everyone following the same bot would all get pinged at the same time. So a team could share a bot account to follow.  I haven't allowed for different followers to have different average times between pings. It's probably easier to have a different bot for different times. 

If a bot account was serving too many followers with pings then rate limiting would kick in at some point, delaying pings. I'm not sure what level that would be. I'll worry about that if it happens. The solution would just be to set up another bot account. 

But even publicising this on the mailing list hasn't resulted in any extra followers. So it's just me and my wife using it for the last year.

**Comment by Insti** [comment]
I started following @cathleenbot & have some feedback, where should I send it?

**Comment by waynemcdougall** [comment]
Hah! Well waynemcdougall@gmail.com will work.

**Comment by mykter** [comment]
I've been working on an Electron (Mac/Windows/Linux) version for a while now, I think it's just about usable at this point. Any feedback or contributions welcome over here: https://github.com/mykter/TagTime-desktop

**Comment by dreeves** [comment]
Exciting! Thanks for building this! I ran into a showstopper answering the first ping and submitted an issue. Hope it's something easy!
